### PR TITLE
ddl: fix the wrong col type when create text(100)

### LINKED ISSUE #27862: Incompatible column type for text
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```
create table t(a text(100));
show create table t;
```

### 2. What did you expect to see? (Required)
```
+-------+--------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                     |
+-------+--------------------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `a` text
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci |
+-------+--------------------------------------------------------------------------------------------------+
1 row in set (0.01 sec)
```
### 3. What did you see instead (Required)
```
+-------+------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                               |
+-------+------------------------------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `a` tinytext DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

In MySQL, the type should be text, not tinytext.

### 4. What is your TiDB version? (Required)
master

### PR DESCRIPTION
**Pull request by sylzd**
### What problem does this PR solve?

Issue Number: close #27862 

Problem Summary:

### What is changed and how it works?

change the lengh compare when adjust blob/text column type.

### Release note 

```release-note
Fix bug that adjusting blob/text column type into a inappropriate type.
```

### CODE REVIEW

**Comment by sylzd** [comment]
M represents the maximum column length in characters both for varchar(M) and text(M), rather than lengh in bytes.  What exactly FieldType.Flen means?

> [NATIONAL] VARCHAR(M) [CHARACTER SET charset_name] [COLLATE collation_name]

A variable-length string. M represents the maximum column length in characters. The range of M is 0 to 65,535. The effective maximum length of a VARCHAR is subject to the maximum row size (65,535 bytes, which is shared among all columns) and the character set used. For example, utf8 characters can require up to three bytes per character, so a VARCHAR column that uses the utf8 character set can be declared to be a maximum of 21,844 characters. See Section 8.4.7, “Limits on Table Column Count and Row Size”.

> TEXT[(M)] [CHARACTER SET charset_name] [COLLATE collation_name]

A TEXT column with a maximum length of 65,535 (216 − 1) characters. The effective maximum length is less if the value contains multibyte characters. Each TEXT value is stored using a 2-byte length prefix that indicates the number of bytes in the value.

An optional length M can be given for this type. If this is done, MySQL creates the column as the smallest TEXT type large enough to hold values M characters long.

File: ddl/db_test.go [hunk 1]
```diff
@@ -7555,3 +7555,18 @@ func (s *testDBSuite8) TestIssue24580(c *C) {
...
+func (s *testDBSuite8) TestCreateTextAdjustLen(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t(a text(100));")
+	tk.MustQuery("show create table t").Check(testutil.RowsWithSep("|", ""+
+		"t CREATE TABLE `t` (\n"+
+		"  `a` text DEFAULT NULL\n"+
+		") ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin"))
+	tk.MustExec("drop table if exists t")
```
Comment (wjhuang2016) [inline ddl/db_test.go:7571]: Please add some tests for `alter table change column`

**Comment by djshow832** [comment]
Please add the release note whenever there's a behavior change, especially when it's a compatibility breaker.

**Review by sylzd — COMMENTED** [review]
> Please add the release note whenever there's a behavior change, especially when it's a compatibility breaker.

I think it is a bug that tinytext cannot hold 100 utf8mb4 charactors.

File: ddl/db_test.go [hunk 2]
```diff
@@ -7555,3 +7555,29 @@ func (s *testDBSuite8) TestIssue24580(c *C) {
...
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t(a text(100));")
+	tk.MustQuery("show create table t").Check(testutil.RowsWithSep("|", ""+
+		"t CREATE TABLE `t` (\n"+
+		"  `a` text DEFAULT NULL\n"+
+		") ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin"))
+	tk.MustExec("alter table t add b text(100);")
+	tk.MustExec("alter table t add c text;")
+	tk.MustExec("alter table t add d text(50);")
```
Comment (wjhuang2016) [inline ddl/db_test.go:7573]: Change d to text(100) and check it's text.

Comment (sylzd) [reply ddl/db_test.go:7573]: Added

**Review by wjhuang2016 — APPROVED** [review]

**Comment by sylzd** [comment]
🤖Review Push Robot 
🤣
/cc @djshow832 @AilinKid

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -555,22 +558,28 @@ func processColumnFlags(col *table.Column) {
...
-			col.FieldType.Tp = mysql.TypeMediumBlob
-		} else if col.FieldType.Flen <= longBlobMaxLength {
-			logutil.BgLogger().Info(fmt.Sprintf("Automatically convert BLOB(%d) to LONGBLOB", col.FieldType.Flen))
-			col.FieldType.Flen = longBlobMaxLength
-			col.FieldType.Tp = mysql.TypeLongBlob
+func adjustBlobTypesFlen(tp *types.FieldType, colCharset string) {
+	cs, err := charset.GetCharsetInfo(colCharset)
+	// when we meet the unsupported charset, we do not adjust.
+	if err != nil {
+		return
```
Comment (djshow832) [inline ddl/ddl_api.go:565]: Typically, we also return the error.

File: ddl/db_test.go [hunk 4]
```diff
@@ -7610,3 +7610,38 @@ func (s *testDBSuite8) TestIssue24580(c *C) {
...
+func (s *testDBSuite8) TestCreateTextAdjustLen(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t(a text(100));")
+	tk.MustQuery("show create table t").Check(testutil.RowsWithSep("|", ""+
+		"t CREATE TABLE `t` (\n"+
+		"  `a` text DEFAULT NULL\n"+
+		") ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin"))
+	tk.MustExec("alter table t add b text(100);")
```
Comment (djshow832) [inline ddl/db_test.go:7626]: Please also test mediumtext and longtext.

**Review by djshow832 — APPROVED** [review]

**Pull request merged**